File: frontend/src/types/session.ts
import type { TagInfo } from "./tag";
import type { WorkflowPhase, WorkflowPhaseStatus } from "./workflow";

export type SessionStatus = "idle" | "running" | "error" | "archived";

export interface CurrentActivity {
  tool: string;
  input?: Record<string, unknown>;
}

export interface SessionInfo {
  id: string;
  status: SessionStatus;
  work_dir: string;
  message_count: number;
  file_changes_count: number;
  claude_session_id?: string;
  created_at?: string;
  allowed_tools?: string;
  system_prompt?: string;
  timeout_seconds?: number;
  workflow_enabled?: boolean;
  workflow_phase?: WorkflowPhase | null;
  workflow_phase_status?: WorkflowPhaseStatus | null;
  permission_mode?: boolean;
  permission_required_tools?: string[];
  name?: string;
  model?: string | null;
  max_turns?: number | null;
  max_budget_usd?: number | null;
  system_prompt_mode?: string | null;
  disallowed_tools?: string | null;
  mcp_server_ids?: string[] | null;
  additional_dirs?: string[] | null;
  fallback_model?: string | null;
  workspace_id?: string | null;
  worktree_name?: string | null;
  parent_session_id?: string | null;
  forked_at_message_id?: number | null;
  tags?: TagInfo[];
  current_activity?: CurrentActivity | null;
}

export interface CreateSessionRequest {
  work_dir?: string | null;
  allowed_tools?: string | null;
  system_prompt?: string | null;
  timeout_seconds?: number | null;
  workflow_enabled?: boolean | null;
  permission_mode?: boolean | null;
  permission_required_tools?: string[] | null;
  model?: string | null;
  max_turns?: number | null;
  max_budget_usd?: number | null;
  system_prompt_mode?: string | null;
  disallowed_tools?: string | null;
  mcp_server_ids?: string[] | null;
  additional_dirs?: string[] | null;
  fallback_model?: string | null;
  worktree_name?: string | null;
  workspace_id?: string | null;
}

export interface UpdateSessionRequest {
  allowed_tools?: string | null;
  system_prompt?: string | null;
  timeout_seconds?: number | null;
  workflow_enabled?: boolean | null;
  permission_mode?: boolean | null;
  permission_required_tools?: string[] | null;
  name?: string | null;
  model?: string | null;
  max_turns?: number | null;
  max_budget_usd?: number | null;
  system_prompt_mode?: string | null;
  disallowed_tools?: string | null;
  mcp_server_ids?: string[] | null;
  additional_dirs?: string[] | null;
  fallback_model?: string | null;
  work_dir?: string | null;
}

export interface ConvertToWorktreeRequest {
  worktree_name: string;
}

export interface SessionStats {
  total_messages: number;
  total_cost: number;
  total_duration_ms: number;
  total_input_tokens: number;
  total_output_tokens: number;
  total_cache_creation_tokens: number;
  total_cache_read_tokens: number;
}

File: backend/app/schemas/session.py
"""세션 관련 Pydantic 요청/응답 스키마."""

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.tag import TagInfo


class CreateSessionRequest(BaseModel):
    work_dir: Optional[str] = None
    allowed_tools: Optional[str] = None
    system_prompt: Optional[str] = None
    timeout_seconds: Optional[int] = Field(None, ge=1, le=7200)
    permission_mode: Optional[bool] = None
    permission_required_tools: Optional[list[str]] = None
    model: Optional[str] = None
    max_turns: Optional[int] = Field(None, ge=1, le=1000)
    max_budget_usd: Optional[float] = Field(None, gt=0)
    system_prompt_mode: Optional[Literal["replace", "append"]] = None
    disallowed_tools: Optional[str] = None
    mcp_server_ids: Optional[list[str]] = None
    additional_dirs: Optional[list[str]] = None
    fallback_model: Optional[str] = None
    worktree_name: Optional[str] = None
    workspace_id: Optional[str] = None
    branch: Optional[str] = None
    workflow_definition_id: Optional[str] = None


class UpdateSessionRequest(BaseModel):
    allowed_tools: Optional[str] = None
    system_prompt: Optional[str] = None
    timeout_seconds: Optional[int] = Field(None, ge=1, le=7200)
    permission_mode: Optional[bool] = None
    permission_required_tools: Optional[list[str]] = None
    name: Optional[str] = None
    model: Optional[str] = None
    max_turns: Optional[int] = Field(None, ge=1, le=1000)
    max_budget_usd: Optional[float] = Field(None, gt=0)
    system_prompt_mode: Optional[Literal["replace", "append"]] = None
    disallowed_tools: Optional[str] = None
    mcp_server_ids: Optional[list[str]] = None
    additional_dirs: Optional[list[str]] = None
    fallback_model: Optional[str] = None
    work_dir: Optional[str] = None
    worktree_name: Optional[str] = None


class CurrentActivity(BaseModel):
    """세션의 현재 활동 정보."""

    tool: str  # "Read", "Write", "Bash", "__thinking__" 등
    input: dict = {}  # {"file_path": "...", "command": "..."} 등


class SessionInfo(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    claude_session_id: Optional[str] = None
    work_dir: str
    status: str
    created_at: datetime
    message_count: int = 0
    file_changes_count: int = 0
    allowed_tools: Optional[str] = None
    system_prompt: Optional[str] = None
    timeout_seconds: Optional[int] = None
    workflow_enabled: bool = True
    workflow_phase: Optional[str] = None
    workflow_phase_status: Optional[str] = None
    permission_mode: bool = False
    permission_required_tools: Optional[list[str]] = None
    name: Optional[str] = None
    model: Optional[str] = None
    max_turns: Optional[int] = None
    max_budget_usd: Optional[float] = None
    system_prompt_mode: str = "replace"
    disallowed_tools: Optional[str] = None
    mcp_server_ids: Optional[list[str]] = None
    additional_dirs: Optional[list[str]] = None
    fallback_model: Optional[str] = None
    workspace_id: Optional[str] = None
    worktree_name: Optional[str] = None
    parent_session_id: Optional[str] = None
    forked_at_message_id: Optional[int] = None
    workflow_definition_id: Optional[str] = None
    tags: list[TagInfo] = []
    current_activity: Optional[CurrentActivity] = None


class ForkSessionRequest(BaseModel):
    """세션 포크 요청. message_id가 None이면 전체 메시지 복사."""

    message_id: Optional[int] = None


class ConvertToWorktreeRequest(BaseModel):
    """기존 세션을 Git 워크트리로 전환 요청."""

    worktree_name: str = Field(
        ..., min_length=1, description="워크트리 이름 (claude -w에 전달)"
    )
